[block-stm] Versioned module storage template

georgemitenkov opened the pull request:
## Description

This PR introduces a new versioned map for module storage, we would cache modules and associated information. Modules can be in two states: pending or published. Pending modules are treated as dependencies, and can be committed (i.e., turned into published). Storage-version modules are always published.

Note that verification and transitive dependency loading will be implemented as part of `view.rs`: on the first call of `set_base_value`, we will pre-load other modules in transitive dependency graph. The actual entry M will be an enum which specifies what information has been cached: deserialized, verified, etc.

## Type of Change
- [x] New feature
- [ ] Bug fix
- [ ] Breaking change
- [ ] Performance improvement
- [ ] Refactoring
- [ ] Dependency update
- [ ] Documentation update
- [ ] Tests

## Which Components or Systems Does This Change Impact?
- [ ] Validator Node
- [ ] Full Node (API, Indexer, etc.)
- [ ] Move/Aptos Virtual Machine
- [ ] Aptos Framework
- [ ] Aptos CLI/SDK
- [ ] Developer Infrastructure
- [ ] Other (specify)

## How Has This Been Tested?
<!--
- Please ensure that the functionality introduced by this change is well tested and verified to work as expected.
- Ensure tests cover both happy and unhappy paths.
- List and link relevant tests.
-->

## Key Areas to Review
<!--
- Identify any critical parts of the code that require special attention or understanding. Explain why these parts are crucial to the functionality or architecture of the project.
- Point out any areas where complex logic has been implemented. Provide a brief explanation of the logic and your approach to make it easier for reviewers to follow.
- Highlight any areas where you are particularly concerned or unsure about the code's impact on the change. This can include potential performance or security issues, or compatibility with existing features.
-->

## Checklist
- [ ] I have read and followed the [CONTRIBUTING](https://github.com/aptos-labs/aptos-core/blob/main/CONTRIBUTING.md) doc
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I identified and added all stakeholders and component owners affected by this change as reviewers
- [ ] I tested both happy and unhappy path of the functionality
- [ ] I have made corresponding changes to the documentation

<!-- Thank you for your contribution! -->

gelash commented on aptos-move/mvhashmap/src/versioned_module_storage.rs:
```diff
@@ -0,0 +1,243 @@
+}
+
+impl<M: Hash> Entry<M> {
+    fn mark_estimate(&mut self) {
```
This is confusing because we are re-using estimate flag for a different purpose here: originally value written at the end of the execution is considered Default = Done, while abort marks them as estimates. Observing the estimate then means a dependency, and they are cleared by the next re-execution.

Here, pending already means an estimate, i.e. that whoever sees pending returns a Dependency, so the only functionality that is missing from the caller point of view, assuming the dependencies should outlive speculation re-executions, is to publish a subset of pending modules, and remove others (during commit) - in that case it might even make sense to record the list of all pending modules for a transaction here (to live across speculative boundaries). We could also choose to clear pending modules alongside other estimates (e.g. from data MVHashMap) after each re-execution, which seems to be the intent here (given the remove API). Worth thinking about. But for this comment's purposes, in any case, marking an Estimate only serves as a way to check an invariant (that whatever we published was from the latest incarnation) - and in that case Done and Estimate are no longer good Flag names.

gelash reviewed (changes requested):
We can add in versioned module storage a list of all module keys that were ever pending for each transaction index (after all, if it became pending, the transaction could actually publish it, right? - we had done the checks in the speculative execution) - so we can wait for commit despite speculative re-executions of the transaction.

We should expose general enough interfaces, e.g.
- add new pending modules: abstractly, all other pending modules of the same index may become stale (up to the implementation). 
- physically remove stale modules for a transaction
- publish: can internally assert a non-stale pending module.

Commit can publish pending modules from the latest (committed) incarnation. Then physically remove stale modules (and ensure all non-stale were published). Then resume transactions that were suspended waiting for its commit (on any pending modules: re-read will observe published or nothing).

The physical remove can also be used to quickly switch to an implementation where pending modules from last incarnation can be cleaned up after each re-execution.

gelash commented on aptos-move/block-executor/src/captured_reads.rs:
```diff
@@ -583,6 +584,30 @@ impl<T: Transaction> CapturedReads<T> {
+                    // TODO(George): Use hash or transaction index to distinguish the reads.
+                    true
+                },
+                Err(Dependency(_)) | Err(Uninitialized) => false,
```
I wonder (even for the data), is it possible to have Uninitialized during validation? maybe it makes sense to use the invariant that validation requires a read to exist, and the read would have initialized the entry? I don't mean adding an unreachable immediately, but maybe start with an error!() @igor-aptos what do you think?

gelash reviewed (commented):
Once we agree on proper interfaces, would also need a mvhashmap prop-test coverage over there ESTIMATEs are inserted, then all execute concurrently, and make sure reads observe correct values in the end. We can do similar with pre-populating with PENDING, and ensuring PUBLISHED are visible in the end.

georgemitenkov commented on aptos-move/mvhashmap/src/versioned_module_storage.rs:
```diff
@@ -0,0 +1,243 @@
+
+#[derive(Debug, Eq, PartialEq)]
+pub enum ModuleReadError {
+    /// There is no entry for this module yet, and it has to be initialized.
```
Yes

georgemitenkov commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/14001?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#14001** <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/14001?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* **#14408** <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/14408?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#14402** <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/14402?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#14397** <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/14397?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#14395** <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/14395?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#14341** <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/14341?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#14206** <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/14206?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#14184** <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/14184?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>: 1 other dependent PR ([#14075](https://github.com/aptos-labs/aptos-core/pull/14075) <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/14075?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>)
* **#14185** <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/14185?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @georgemitenkov and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

gelash commented on aptos-move/block-executor/src/captured_reads.rs:
```diff
@@ -583,6 +625,25 @@ impl<T: Transaction> CapturedReads<T> {
+        // and fetching it returned an error, the error is not recorded. Hence, it is safe here
+        // to simply treat such errors as a non-existent value.
+        self.module_storage_reads.iter().all(|(k, previous_read)| {
+            let result = module_storage.get(k, ShiftedTxnIndex::new(idx_to_validate));
```
I think it's better to call with index and shift it inside implementation (also convention for existing interfaces afair)

georgemitenkov replied:
done

gelash commented on aptos-move/block-executor/src/captured_reads.rs:
```diff
@@ -583,6 +625,25 @@ impl<T: Transaction> CapturedReads<T> {
+        // to simply treat such errors as a non-existent value.
+        self.module_storage_reads.iter().all(|(k, previous_read)| {
+            let result = module_storage.get(k, ShiftedTxnIndex::new(idx_to_validate));
+            let new_read = ModuleStorageRead::new(&result);
```
nit: maybe we can implement comparison directly, without having to construct a new ModuleStorageRead?

georgemitenkov replied:
Yes, I reused read in both MV data structure and here

gelash commented on aptos-move/block-executor/src/view.rs:
```diff
@@ -1534,6 +1541,21 @@ impl<'a, T: Transaction, S: TStateView<Key = T::Key>, X: Executable> TModuleView
+        // Enforce feature gating V2 loader implementation: TModuleView is no longer used in
+        // V2 interfaces because we implement storage traits directly. Use a debug assert to
+        // panic in tests, adn invariant violation for non-debug builds.
+        if self.runtime_environment.vm_config().use_loader_v2 {
```
if we only need use_loader_v2, then can we pass that to view constructor (instead of the whole runtime environment)?

georgemitenkov replied:
We need the environment to run verification and deserialization, it is mostly used in `modules_and_scripts.rs`

gelash commented on aptos-move/block-executor/src/view.rs:
```diff
@@ -1556,10 +1578,13 @@ impl<'a, T: Transaction, S: TStateView<Key = T::Key>, X: Executable> TModuleView
-                )
+                state
+                    .unsync_map
+                    .fetch_module_for_loader_v2(state_key)
```
hm, above errors if we are using loader v2, but here the API is renamed to v2? in other words, is it just that the API was renamed and is used in the old flow, or do we actually go through this flow in sequential case even for v2? The latter seems bad, but I think it's the former, in which case maybe we should adjust the name to indicate it works in both cases?

georgemitenkov replied:
Good catch, I renamed incorrectly, should be `.._loader_v1`.  Fixed.

gelash commented on aptos-move/block-executor/src/view.rs:
```diff
@@ -1534,6 +1541,21 @@ impl<'a, T: Transaction, S: TStateView<Key = T::Key>, X: Executable> TModuleView
+                PartialVMError::new(StatusCode::UNKNOWN_INVARIANT_VIOLATION_ERROR)
+                    .with_message(msg),
+            );
+            debug_assert!(err.is_ok());
```
how can err be ok?

georgemitenkov replied:
Used a debug assert to panic in tests.  But otherwise in prod, we cannot, so it is some sort of gating where we just return an invariant violation. Not ideal, maybe we just debug assert it?

This actually caused some panics in prop tests where I have not implement module storage yet, so it was useful.

gelash commented on aptos-move/block-executor/src/modules_and_scripts.rs:
```diff
@@ -0,0 +1,300 @@
+            },
+            ViewState::Unsync(state) => {
+                state.read_set.borrow_mut().module_reads.insert(key.clone());
+                Ok(match state.unsync_map.fetch_module(key) {
```
hm, it this flipped? shouldn't here be v2 and in view.rs the normal fetch_module?

georgemitenkov replied:
I flipped implementations, this uses the new flow, old API is renamed to `if_using_...`

gelash commented on aptos-move/block-executor/src/captured_reads.rs:
```diff
@@ -298,9 +325,14 @@ impl DelayedFieldRead {
-    // TODO: implement a general functionality once the fallback is removed.
+    /// Captured module reads if V1 loader is used.
     pub(crate) module_reads: Vec<T::Key>,
+    /// Captured module reads if V2 loader is used. Here, using a vector is important for
```
If it's not too troublesome, I'd much rather avoid this comment and have a hashmap that records the ModuleStorageReadResults directly and acts as a cache - i.e. in modules_and_scripts, we would first check if it's recorded and return, otherwise, we would fetch and record. We can project to ModuleStorageRead (or the new name) for validation purposes. This way CapturedReads acts as a single-threaded cache/consistency layer for the executing session, which is beneficial for reducing the surface of speculative inconsistencies, as well as generally good practice for performance (and consistent with what we do in captured reads for resources, etc)

georgemitenkov replied:
Good point, will do this

georgemitenkov replied:
Done

gelash commented on third_party/move/move-vm/runtime/src/storage/dummy.rs:
```diff
@@ -1,103 +1,7 @@
-}
-
 // TODO(loader_v2): Temporary infra to enable loader V2 to test & run things e2e locally, remove.
 pub fn use_loader_v2_based_on_env() -> bool {
```
hm, let's also ship this somewhere (where it's used) and get rid of the file?

georgemitenkov replied:
Did this in one of the later PRs: #14495

gelash commented on types/src/executable.rs:
```diff
@@ -15,12 +15,15 @@ pub enum ExecutableDescriptor {
 pub trait ModulePath {
-    // TODO(George):
+    // TODO(loader_v2):
     //   Improve this in the future, right now all writes use state keys
```
supernit / OCD : comment can fit in 2 lines :)

georgemitenkov replied:
I keep doing 80 lines and forget that our default is 100😄. FIxed.

gelash commented on aptos-move/block-executor/src/modules_and_scripts.rs:
```diff
@@ -0,0 +1,300 @@
+        &self,
+        state_key: &StateKey,
+    ) -> PartialVMResult<Option<usize>> {
+        // TODO(loader_v2): A very ugly way of converting state keys into generic types.
```
🙈

georgemitenkov replied:
I think we need to have AptosModuleStorage generic, and even for module storage: all we need is a generic key which is constructible from address and name. Then we can use ModuleId in MoveVM, StateKey in Aptos, and avpid nested maps. And in future, some index-like compression? Wdyt?

georgemitenkov replied:
Because the trait is not generic, so I guess it should be

georgemitenkov replied:
Fixed in the main PR.

georgemitenkov commented on aptos-move/block-executor/src/captured_reads.rs:
```diff
@@ -286,6 +287,32 @@ impl DelayedFieldRead {
+#[derive(Clone, Eq, PartialEq, Debug)]
+pub(crate) enum ModuleStorageRead {
+    /// Whether the read module existed or not. Modules that are pending to be
+    /// published and committed yield this as well.
```
Removed this in favour of `ModuleStorageReadResult` as they are essentially the same thing

georgemitenkov commented on aptos-move/block-executor/src/view.rs:
```diff
@@ -2441,8 +2466,13 @@ mod test {
         let base_view = MockStateView::new(HashMap::new());
         let start_counter = 5;
+
+        // FIXME
```
Fixed in subsequent PR

georgemitenkov commented on aptos-move/aptos-vm-types/src/environment.rs:
```diff
@@ -5,15 +5,37 @@ use aptos_gas_schedule::{gas_feature_versions::RELEASE_V1_15, AptosGasParameters
-use std::sync::Arc;
+use std::sync::{Arc, OnceLock};
+
+// TODO(loader_v2): Remove this, it is for tests only.
+static ENV: OnceLock<Arc<RuntimeEnvironment>> = OnceLock::new();
```
Yes, not for prod, a temporary hack to get things running end to end and test. The proper solution is to read it in block executor from the state. I will add a PR for that once everything else seems reasonable and good

georgemitenkov replied:
Fixed in the main PR.

georgemitenkov commented on aptos-move/aptos-vm/src/errors.rs:
```diff
@@ -138,6 +138,7 @@ pub fn convert_prologue_error(
                     let err_msg = format!("[aptos_vm] Unexpected prologue Move abort: {:?}::{:?} (Category: {:?} Reason: {:?})",
                     location, code, category, reason);
                     speculative_error!(log_context, err_msg.clone());
+                    println!("B: {:?}", err_msg);
```
Oops

georgemitenkov commented on aptos-move/block-executor/src/modules_and_scripts.rs:
```diff
@@ -0,0 +1,325 @@
+        &self,
+        _serialized_script: &[u8],
+    ) -> VMResult<Arc<CompiledScript>> {
+        // TODO(loader_v2): Support scripts.
```
It will cache scripts with the same code (so no need to reverify) within the block. The next PR in the stack adds some basic support there

georgemitenkov commented on aptos-move/block-executor/src/executor.rs:
```diff
@@ -493,6 +542,21 @@ where
                     ),
                 )?;
 
+                // Make sure we publish modules here before we wake up subsequent dependent
+                // transactions and decrease the validation index.
+                if runtime_environment.vm_config().use_loader_v2 {
+                    if let Some(module_write_set) = last_input_output.module_write_set(txn_idx) {
+                        executed_at_commit = true;
+                        for (key, v) in module_write_set.into_iter() {
+                            let entry =
+                                ModuleStorageEntry::from_transaction_write(runtime_environment, v)?;
+                            versioned_cache
+                                .module_storage()
+                                .write_published(&key, txn_idx, entry);
+                        }
+                    }
+                }
+
                 scheduler.finish_execution_during_commit(txn_idx)?;
```
Will address this as a part of a general cleanup, added a TODO

georgemitenkov commented:
@gelash let's iterate more on this PR, but I have addresses comments otherwise and will merge it so that I can test & have a single big PR on top of which I will push 1) commits which make reviewing easier and 2) PRs which are more major changes
